ls1intum/Artemis
# Fix Usage of Undefined Exam Exercise Dates

### Checklist

### Motivation and Context
Fix #3821 and related issues. The problem is, that we use the `Exercise` object's dates in some places that need to handle both exam and course exercises and the dates (release, due, assessment due) are undefined (`null` in Java) for exam exercises.

### Description
- *Temporary: added assertions to methods in `Exercise` that must not be used for exam exercises.*
- Fix multiple issues where the presence of exam dates was assumed, but exam exercises have no dates.
- Make the WebSocket respect exam exercise results.
- Add many tests for special exam exercise checks.

#### TODO
- [x] Investigate all failing tests
- [x] Look for more date related problems - future work ??
- [x] ~Redefine and combine some methods in exercise, esp. `isVisibleToStudents` and `isReleased`~
- [x] ~Adjust `validateDates` to make sure, the dates are not defined for exam exercises?~

### Steps for Testing
- Check that #3821 cannot be reproduced.
  
  With an exercise with (almost) all test cases marked after due date:
  ![image](https://user-images.githubusercontent.com/11130248/128650771-70178599-6139-4798-b296-c8a23bd72619.png)
  Now all test cases are shown, independent of the set build after due date.

- You should also check that you can still work in the code editor on programming exercises and get your results when the build is done *without page reload* (meaning that you still get notified by the server).

### Test Coverage
- **WebsocketMessagingService**
  ![image](https://user-images.githubusercontent.com/11130248/128650000-e1aa71d5-4515-4c02-9bc5-8d018b917012.png)
- **ExamDateService**
  ![image](https://user-images.githubusercontent.com/11130248/128650002-947f77e3-d902-4fbd-ab82-c6cd7511f6fe.png)
- **QuizSubmissionResource**
  ![image](https://user-images.githubusercontent.com/11130248/128650008-269823bf-618f-4e12-b15d-7f1f2fbcdce0.png)
